x/icacallbacks/icacallbacks.go

```diff
@@ -0,0 +1,84 @@
+package icacallbacks
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	ibctransfertypes "github.com/cosmos/ibc-go/v5/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v5/modules/core/04-channel/types"
+	"github.com/gogo/protobuf/proto"
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/Stride-Labs/stride/v4/x/icacallbacks/types"
+)
+
+// Parses ICA tx responses and returns a list of each serialized response
+// The format of the raw ack differs depending on which version of ibc-go is used
+// For v4 and prior, the message responses are stored under the `Data` attribute of TxMsgData
+// For v5 and later, the message responses are stored under the `MsgResponse` attribute of TxMsgdata
+func ParseTxMsgData(acknowledgementResult []byte) ([][]byte, error) {
+	txMsgData := &sdk.TxMsgData{}
+	if err := proto.Unmarshal(acknowledgementResult, txMsgData); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-27 tx message data: %s", err.Error())
+	}
+
+	// Unpack all the message responses based on the sdk version (determined from the length of txMsgData.Data)
+	switch len(txMsgData.Data) {
+	case 0:
+		// for SDK 0.46 and above
+		msgResponses := make([][]byte, len(txMsgData.MsgResponses))
+		for i, msgResponse := range txMsgData.MsgResponses {
+			msgResponses[i] = msgResponse.GetValue()
+		}
+		return msgResponses, nil
+	default:
+		// for SDK 0.45 and below
+		var msgResponses = make([][]byte, len(txMsgData.Data))
+		for i, msgData := range txMsgData.Data {
+			msgResponses[i] = msgData.Data
+		}
+		return msgResponses, nil
+	}
+}
+
+// UnpackAcknowledgementResponse unmarshals IBC Acknowledgements, determines the status of the acknowledgement (success or failure)
+//     and, if applicable, assembles the message responses
+// ICA transactions have associated messages responses. IBC transfer do not.
+// With ICA transactions, the schema of the response differs depending on the version of ibc-go used,
+//     however, this function unifies the format into a common response (a slice of byte arrays)
+func UnpackAcknowledgementResponse(ctx sdk.Context, logger log.Logger, ack []byte, isICA bool) (*types.AcknowledgementResponse, error) {
```

Comment:
Not needed here, in the future we might want to encode the ack type as an enum (ICA, Transfer, etc.) so that we can support more IBC apps

Reply:
That's a great idea, much cleaner! I'll actually make a quick PR for that soon